src/ring_theory/ideal/nc_jacobson_radical.lean

```diff
@@ -0,0 +1,250 @@
+
+/-- An element of a monoid is a unit
+if and only if it has both a left inverse and a right inverse. -/
+theorem is_unit_iff_has_left_inv_right_inv :
+  is_unit a ↔ has_left_inv a ∧ has_right_inv a :=
+⟨λ h, ⟨h.exists_left_inv, h.exists_right_inv⟩,
+  λ ⟨⟨x, hxa⟩, ⟨y, hay⟩⟩, ⟨⟨a, x, by { convert hay,
+  rw [←mul_one x, ←hay, ←mul_assoc, hxa, one_mul] }, hxa⟩, rfl⟩⟩
+
+/-- An element of a monoid is a unit
+if it has a left inverse which also has a left inverse. -/
+theorem is_unit_of_has_left_inv_of_has_left_inv
+  (h₁ : b * a = 1) (h₂ : has_left_inv b) : is_unit a :=
+let ⟨c, hcb⟩ := h₂ in ⟨⟨a, b, by { convert hcb,
+  rw [←one_mul a, ←hcb, mul_assoc, h₁, mul_one] }, h₁⟩, rfl⟩
+
+end monoid
+
+section ring
+variables {R : Type u} {a : R} {b : R} [ring R]
+/-! ### Some lemmas on inverses in rings -/
+
+lemma has_left_inv.one_add_mul_swap
+  (h : has_left_inv (1 + a * b)) : has_left_inv (1 + b * a) :=
+begin
+  obtain ⟨u, hu⟩ := h,
+  existsi 1 - b * u * a,
+  calc (1 - b * u * a) * (1 + b * a)
+        = 1 + b * a - b * (u * (1 + a * b)) * a : by noncomm_ring
+    ... = 1 : by rw [hu, mul_one, add_sub_cancel],
+end
+
+lemma has_right_inv.one_add_mul_swap
+  (h : has_right_inv (1 + a * b)) : has_right_inv (1 + b * a) :=
+begin
+  obtain ⟨u, hu⟩ := h,
+  existsi 1 - b * u * a,
+  calc (1 + b * a) * (1 - b * u * a)
+        = 1 + b * a - b * ((1 + a * b ) * u) * a : by noncomm_ring
+    ... = 1 : by rw [hu, mul_one, add_sub_cancel],
+end
+
+lemma is_unit.one_add_mul_swap
+  (h : is_unit (1 + a * b)) : is_unit (1 + b * a) :=
+begin
+  rw is_unit_iff_has_left_inv_right_inv at *,
+  obtain ⟨h₁, h₂⟩ := h,
+  exact ⟨h₁.one_add_mul_swap, h₂.one_add_mul_swap⟩,
+end
+
+end ring
+
+namespace ideal
+section nc_jacobson_radical
+variables {R : Type u} [ring R]
+/-! ### Jacobson radical of a ring -/
+
+/--
+For a ring `R`, `nc_jacobson R` is the Jacobson radical of `R`, that is,
+the intersection of all maximal left ideals of of `R`. Note that we use left ideals.
+-/
+def nc_jacobson (R : Type u) [ring R] : ideal R := jacobson ⊥
+
+lemma nc_jacobson_def : nc_jacobson R = Inf {I : ideal R | I.is_maximal } :=
+congr_arg Inf $ by simp
+
+lemma has_left_inv_iff_span_top {x : R} :
+  has_left_inv x ↔ span ({x} : set R) = ⊤ :=
+begin
+  split,
+  { rintro ⟨a, hax⟩,
+    apply eq_top_of_unit_mem _ x a _ hax,
+    apply submodule.mem_span_singleton_self },
+  { intro h,
+    have : (1 : R) ∈ span ({x} : set R) := by { rw h, exact submodule.mem_top },
+    exact (mem_span_singleton').mp this },
+end
+
+lemma not_has_left_inv_iff_mem_maximal {x : R} :
+  ¬has_left_inv x ↔ ∃ I : ideal R, I.is_maximal ∧ x ∈ I :=
+begin
+  rw has_left_inv_iff_span_top,
+  split,
+  { intro hx,
+    obtain ⟨I, hImax, hxI⟩ := exists_le_maximal _ hx,
+    exact ⟨I, hImax, by {apply hxI, apply submodule.mem_span_singleton_self}⟩ },
+  { rintro ⟨I, hImax, hxI⟩ hcontra,
+    refine hImax.ne_top _,
+    rwa [eq_top_iff, ←hcontra, span_le, set.singleton_subset_iff] },
+end
```

Comment:
Personally, I would move this part up, as it does not depend on the definition of the Jacobson radical. This would also make it easier to split it up into its own file when the file grows in the future.

Reply:
Since similar facts about ideals are written in `ring_theory.ideal.basic`, I moved these two lemmas to `ring_theory.ideal.basic`.